elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/filter/visitor/SplitFilterExpressionVisitor.java

```diff
@@ -0,0 +1,247 @@
+import com.yahoo.elide.core.filter.FilterPredicate;
+import com.yahoo.elide.core.filter.expression.AndFilterExpression;
+import com.yahoo.elide.core.filter.expression.FilterExpression;
+import com.yahoo.elide.core.filter.expression.FilterExpressionVisitor;
+import com.yahoo.elide.core.filter.expression.NotFilterExpression;
+import com.yahoo.elide.core.filter.expression.OrFilterExpression;
+import com.yahoo.elide.datastores.aggregation.annotation.MetricAggregation;
+import com.yahoo.elide.parsers.expression.FilterExpressionNormalizationVisitor;
+
+import org.apache.commons.lang3.tuple.Pair;
+
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.lang.annotation.Annotation;
+import java.util.Objects;
+
+/**
+ * {@link SplitFilterExpressionVisitor} splits the {@link FilterExpression} into a {@code WHERE} expression and a
+ * {@code Having} expression.
+ * <p>
+ * {@link SplitFilterExpressionVisitor} is leveraged by the AggregationStore to construct the JPQL query.
+ * {@link FilterExpression} for AggregationDataStore must be split into those that apply to metric aggregations
+ * ({@code HAVING} clauses) and those that apply to dimensions ({@code WHERE} clauses), although only a single
+ * {@link FilterExpression} is passed to the datastore in each query. The split groups {@code HAVING} clauses and
+ * {@code WHERE} clauses separately. For example:
+ * <pre>
+ * +-------------------+-----------------------------+------------------------+
+ * |    Expression     |             SQL             | WHERE-clause Promotion |
+ * +-------------------+-----------------------------+------------------------+
+ * | H1 AND W1 AND H2  | WHERE W1 HAVING (H1 AND H2) | No                     |
+ * | (H1 OR H2) AND W1 | WHERE W1 HAVING (H1 OR H2)  | No                     |
+ * | H1 OR W1          | HAVING (H1 OR W1)           | Yes                    |
+ * | (W1 AND H1) OR W2 | HAVING ((W1 AND H1) OR W2)  | Yes                    |
+ * +-------------------+-----------------------------+------------------------+
+ * </pre>
+ * Note that {@link SplitFilterExpressionVisitor} might incur more-than-expected network I/O in the case of WHERE-clause
+ * promotion.
+ * <p>
+ * {@link SplitFilterExpressionVisitor} splits by storing {@code WHERE} and {@code HAVING} clauses in
+ * {@link Pair#getLeft() left} and {@link Pair#getRight() right} of a {@link Pair}, respectively. For example:
+ * <pre>
+ * {@code
+ * Pair<FilterExpression, FilterExpression> filterPair = filterExpression.accept(splitFilterExpressionVisitor);
+ *
+ * FilterExpression whereClauseFilter = filterPair.getLeft();
+ * FilterExpression havingClauseFilter = filterPair.getRight();
+ * }
+ * </pre>
+ * {@link SplitFilterExpressionVisitor} is thread-safe and can be accessed by multiple threads at the same time.
+ */
+@Slf4j
+public class SplitFilterExpressionVisitor implements FilterExpressionVisitor<Pair<FilterExpression, FilterExpression>> {
```

Comment:
We can create a class WhereHaving to store where and having clauses instead of using Pair.

Reply:
:+1: on having a WhereHaving class.